OpenRA.Mods.Cnc/FileSystem/MegFile.cs

```diff
@@ -0,0 +1,206 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using OpenRA.FileSystem;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Cnc.FileSystem
+{
+	/// <summary>
+	/// This class supports loading of the .meg V3 file format with the reference
+	/// documentation here https://modtools.petrolution.net/docs/MegFileFormat
+	/// There are 3 variants, and encrypted support that have not been implemented yet
+	/// </summary>
+	public class MegV3Loader : IPackageLoader
+	{
+		public bool TryParsePackage(Stream s, string filename, OpenRA.FileSystem.FileSystem context, out IReadOnlyPackage package)
+		{
+			var reader = new BinaryReader(s);
+			var id1 = reader.ReadUInt32();
+			var id2 = reader.ReadUInt32();
+
+			if (!id1.Equals(HeaderId1) || !id2.Equals(HeaderId2))
+			{
+				package = null;
+				return false;
+			}
+
+			package = new MegFile(s, filename);
+			return true;
+		}
+
+		internal const int HeaderNumberOfFileNamesOffset = 0xC;
+		internal const int HeaderNumberOfFilesOffset = 0x10;
+		internal const int HeaderFileTableSizeOffset = 0x14;
+		internal const int HeaderFileNameTableStartOffset = 0x18;
+
+		internal const uint HeaderId1 = 0xffffffff;
+		internal const uint HeaderId2 = 0x3F7D70A4;
+
+		public sealed class MegFile : IReadOnlyPackage
+		{
+			private readonly Stream s;
+
+			private readonly List<string> contents = new List<string>();
+
+			private readonly List<MegFileContentReference> fileData = new List<MegFileContentReference>();
+			private readonly string name;
+
+			internal MegFile(FileStream s)
+			{
+				this.s = s;
+
+				var reader = new BinaryReader(s);
+
+				var id1 = reader.ReadUInt32();
+				var id2 = reader.ReadUInt32();
+
+				if (!id1.Equals(HeaderId1) || !id2.Equals(HeaderId2))
```

Comment:
`if (id1 != HeaderId1 || id2 != HeaderId2)` this isn't java :smile:

Reply:
There are cases where this is necessary, but certainly not here. I didn't even know this was a thing for Java (I don't code in it).